Do not merge - code review of the entire code

rklaehn opened the pull request.

rklaehn committed:
Add second representation of a TagSetSet

rklaehn committed:
More tests

rklaehn committed:
WIP handle giant case

rklaehn committed:
WIP

rklaehn committed:
WIP

rklaehn committed:
WIP more tests

rklaehn committed:
Add more tests and stuff

Still something weird - I don't think the property based tests test very much...

rklaehn committed:
Better display

rklaehn committed:
Moar tests

rklaehn committed:
Restructure

rklaehn committed:
Get rid of some dependencies

rklaehn committed:
Add realistic example

rklaehn committed:
Nicer Display

rklaehn committed:
Use an Arc in the bench...

rklaehn committed:
Some tools for zero allocation decoding

rklaehn committed:
Change crate structure and make it a lib

rklaehn committed:
Add sparse bench

rklaehn committed:
Add example for running cargo flamegraph

rklaehn committed:
Add size tests

rklaehn committed:
No allocations in from_fallible_fn

rklaehn committed:
Move read_seq and friends to util

rklaehn committed:
Use where

...just nicer if you have tons of constraints

rklaehn committed:
Use where

...just nicer if you have tons of constraints

rklaehn committed:
Add CI

rklaehn committed:
Make clippy happy

rklaehn committed:
Merge remote-tracking branch 'actyx/master'

# Conflicts:
#	src/util.rs

rklaehn committed:
Add get and make a few methods non-test

benjamin-actyx commented on src/lib.rs:
```diff
@@ -0,0 +1,673 @@
+            .iter()
+            .map(|tag| self.tags.get(tag).cloned())
+            .collect::<Box<_>>();
+        dnf_query0(&self.sets, dnf, &translate, result);
```
The way I understand this, the block will always be scanned through, even if there is no overlap in tags (?)

rklaehn replied:
Yes, I guess I could add a shortcut. But I think what it will do is basically just set everything to false.

```
            let dnf = DenseBitmap::new(
                dnf.sets
                    .iter()
                    .filter_map(|row| {
                        mask_from_bits_iter(
                            row.map(|index| translate[index as usize].unwrap_or(128)),
                        )
                        .ok()
                    })
                    .collect(),
            );
            for (mask, value) in index.iter().zip(result.iter_mut()) {
                *value = *value && { dnf.iter().any(move |query| is_subset(*query, *mask)) }
            }
```

dnf.iter() is empty, any return false, so this ends up being:

```
            for (mask, value) in index.iter().zip(result.iter_mut()) {
                *value = *value && false
            }
```

Not sure if that warrants a special case.

benjamin-actyx replied:
Ah right, the output must still be modified.
Still might warrant a special case to get around loading all the masks in `index` – they’re inside a `Vec` so they actually may be fragmented in memory.

benjamin-actyx commented on src/lib.rs:
```diff
@@ -0,0 +1,673 @@
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct DnfQuery<T: Tag> {
+    tags: Vec<T>,
+    sets: Bitmap,
```
I haven’t understood how it helps to keep this as `Bitmap`. Wouldn’t the translation process be less indirect if these were just sets of tags?

rklaehn replied:
In both cases you want to iterate over the indexes. Iterating over the one bits of a u128 should be pretty fast. You are probably right that it does not make a giant difference though...

benjamin-actyx commented on src/size_tests.rs:
```diff
@@ -0,0 +1,72 @@
+/// less than 128 distinct tags - dense index
+const DENSE: &[(&str, usize)] = &[("article", 30), ("sku", 40), ("location", 40)];
+// more than 128 distinct tags - sparse index
+// const SPARSE: &[(&str, usize)] = &[("article", 30), ("sku", 40), ("location", 100)];
```
A good real-world example for sparse would be the initial importer blocks we create. They will have hundreds of `ProcessCreated` with a different process-id-tag each.

benjamin-actyx commented on src/lib.rs:
```diff
@@ -0,0 +1,673 @@
+                    .iter()
+                    .filter_map(|row| {
+                        mask_from_bits_iter(
+                            row.map(|index| translate[index as usize].unwrap_or(128)),
```
Why 128?

rklaehn replied:
128 causes mask_from_bits_iter to return an error, which means that the row will be skipped due to the .ok() below.

benjamin-actyx replied:
So would this also work?
`row.map(|index| translate[index as usize]).and_then(mask_from_bits_iter).ok()`
I find the indirection via that 128 confusing

rklaehn replied:
I agree it is less than optimal, but I played around with it a bit and could not come up with something better.

The problem is that mask_from_bits_iter takes an iterator of indexes, not an iterator of options. So a shortcut to skip the row is to just pass in an index that will make it fail.

Until we come up with something better, I added a longer comment.

benjamin-actyx commented on src/lib.rs:
```diff
@@ -0,0 +1,673 @@
+        assert_eq!(&matches("ab,ac,bc", " ax"), "000");
+        assert_eq!(&matches("ab,ac,bc", "  c"), "011");
+        assert_eq!(&matches("ab,bc,cd", "a|d"), "101");
+    }
```
These are missing tests, right?
- sparse version
- disjunct tag sets
- matches that are not within the first 8 bit

rklaehn replied:
yes, this was mostly me playing around. Could definitely need some more.

benjamin-actyx commented on src/bitmap.rs:
```diff
@@ -0,0 +1,470 @@
+// set for the sparse case
+pub(crate) type IndexSet = VecSet<[u32; 4]>;
+// mask for the dense case
+pub(crate) type IndexMask = u128;
```
This crate looks kind of nice? https://docs.rs/bitmaps/3.1.0/bitmaps/
Using it, the bit-twiddling in this file would be replaced by somewhat more expressive function calls like `set(bit, true)`.
____
There is also Roaring bitmaps, which includes a compression scheme and stable serialization format:
https://github.com/RoaringBitmap/roaring-rs 
It won most of the union and intersection benchmarks I showed yesterday: 
https://people.csail.mit.edu/jshun/6886-s20/lectures/lecture19-1.pdf (Slide 31+)

However, Roaring is probably a bit pointless for 128-bit maps. From what I understand, its strength lies in supporting really huge bitmaps while retaining a small memory footprint. So it would rather serve as a replacement for both sparse and dense cases combined.
Anyways, might be interesting to benchmark.

benjamin-actyx replied:
Could also be an interesting replacement for `matches: &mut [bool]`

rklaehn replied:
`matches: &mut [bool]` used to be some bitmap crate. I found it to be a pain in the ass and not worth it.

rklaehn replied:
I think roaring would make more sense for the transposed version that you suggested, so it has more to work with.

benjamin-actyx commented on src/lib.rs:
```diff
@@ -0,0 +1,673 @@
+pub struct TagIndex<T: Tag> {
+    /// efficiently encoded distinct tags
+    tags: TagSetSet<T>,
+    /// tag offset for each event
+    events: Vec<u32>,
```
I don’t understand this. From the docs there are not "offsets" but bitmaps – but why is `u32` large enough in this case?

rklaehn replied:
These are just offsets into the bitmaps of the above TagSetSet. So u32 should be big enough. I was even thinking of limiting the number of events per block to u16 to save some in memory representation space.

benjamin-actyx replied:
Aaah, I get it. So it is `tag set offset for each event`
meaning
`the tags for event 5 can be found via tags[events[4]]`.

rklaehn replied:
yes. Not sure how much this is worth. Especially in the dense case, the subset ops are already so fast that this additional level of indirection might not buy much. But in the sparse case it is worth something.

rklaehn committed:
Remove delta-encoding and sort by tags in builder

rklaehn committed:
Remove TagSetSet

...now that we canonicalize, it is no longer needed.

rklaehn committed:
Make it easy to switch the width of a dense bitmap

rklaehn committed:
Switch to 64 bits for the dense case

The sparse case is also pretty fast, so a big benefit of the dense case
is the more compact representation. Using u64 instead of u128 makes this
advantage of the dense case bigger.

rklaehn committed:
Release 0.1.1

rklaehn committed:
Remove the TagSet type alias, making the vec_collections dependency invisible

This requires changing some return types.

rklaehn committed:
Release 0.1.2

rklaehn committed:
Add validation on decode...

so we don't end up with queries or indexes that panic as soon as you look at them.

rklaehn committed:
Builder now just builds one thing, so rename dnf_query to result

rklaehn committed:
Release 0.1.3

rklaehn committed:
Add comment about purpose of Tag trait

rklaehn committed:
Release 0.2.0

wngr committed:
dont depend explicitly on libipld-cbor

wngr committed:
update deps

wngr committed:
increment version

wngr committed:
some leftovers

wngr committed:
Merge pull request #2 from Actyx/ow/deps

Simplify dependencies

rkuhn committed:
update dependencies and release 0.3.0
